src/net/tls.cc

```diff
@@ -1533,25 +1533,110 @@ class session : public enable_lw_shared_from_this<session> {
         result_t dn = extract_dn_information();
         return make_ready_future<result_t>(std::move(dn));
     }
+    future<std::vector<subject_alt_name>> get_alt_name_information(std::unordered_set<subject_alt_name_type> types) {
+        using result_t = std::vector<subject_alt_name>;
+
+        if (_error) {
+            return make_exception_future<result_t>(_error);
+        }
+        if (_shutdown) {
+            return make_exception_future<result_t>(std::system_error(ENOTCONN, std::system_category()));
+        }
+        if (!_connected) {
+            return handshake().then([this, types = std::move(types)]() mutable {
+               return get_alt_name_information(std::move(types));
+            });
+        }
+
+        auto gris = extract_dn_information();
+        auto peer = get_peer_certificate();
+        if (!peer) {
+            return  make_ready_future<result_t>();
+        }
+
+        result_t res;
+
+    	for (auto i = 0u; ; i++) {
+            size_t size = 0;
+
+            auto err = gnutls_x509_crt_get_subject_alt_name(peer.get(), i, nullptr, &size, nullptr);
+
+            if (err == GNUTLS_E_REQUESTED_DATA_NOT_AVAILABLE) {
+                break;
+            }
+            if (err != GNUTLS_E_SHORT_MEMORY_BUFFER) {
+                gtls_chk(err); // will throw
+            }
+            sstring buf;
+            buf.resize(size);
+
+            err = gnutls_x509_crt_get_subject_alt_name(peer.get(), i, buf.data(), &size, nullptr);
+            if (err < 0) {
+                gtls_chk(err); // will throw
```

Comment:
Better to wrap the whole function body in futurize_invoke so the exception is made part of the future, leading to fewer surprises.

Reply:
Fixed.